File: server/src/graphql/resolvers/mutation.ts
import { randomUUID } from "crypto";

import { createLibrary, deleteLibrary, updateLibrary } from "../../db/queries/libraries.js";
import { insertPlaybackSession } from "../../db/queries/playbackHistory.js";
import { setSetting } from "../../db/queries/userSettings.js";
import { deleteVideoMetadata, upsertVideoMetadata } from "../../db/queries/videoMetadata.js";
import { getVideoById } from "../../db/queries/videos.js";
import {
  addWatchlistItem,
  removeWatchlistItem,
  updateWatchlistProgress,
} from "../../db/queries/watchlist.js";
import type { GQLContext } from "../../routes/graphql.js";
import { startTranscodeJob } from "../../services/chunker.js";
import { scanLibraries } from "../../services/libraryScanner.js";
import { fetchOmdbById, isOmdbConfigured } from "../../services/omdbService.js";
import type { MediaType, VideoMetadataRow } from "../../types.js";
import { gqlMediaTypeToInternal, gqlResolutionToInternal } from "../mappers.js";
import {
  type GQLLibrary,
  type GQLPlaybackError,
  type GQLPlaybackSession,
  type GQLTranscodeJob,
  type GQLVideo,
  type GQLWatchlistItem,
  presentJob,
  presentLibrary,
  presentPlaybackError,
  presentPlaybackSession,
  presentVideo,
  presentWatchlistItem,
} from "../presenters.js";
import { fromGlobalId } from "../relay.js";

export const mutationResolvers = {
  /**
   * Discriminates union members by the `__typename` set on the presenter
   * output. Required by graphql-tools because the union has no `interface`
   * field that yoga could auto-resolve from. Lives at the top level (not
   * inside `Mutation`) because `StartTranscodeResult` is a type, not a
   * Mutation field.
   */
  StartTranscodeResult: {
    __resolveType(obj: GQLTranscodeJob | GQLPlaybackError): string {
      return obj.__typename;
    },
  },

  Mutation: {
    async scanLibraries(): Promise<GQLLibrary[]> {
      const rows = await scanLibraries();
      return rows.map(presentLibrary);
    },

    async startTranscode(
      _: unknown,
      {
        videoId,
        resolution,
        startTimeSeconds,
        endTimeSeconds,
      }: {
        videoId: string;
        resolution: string;
        startTimeSeconds?: number;
        endTimeSeconds?: number;
      },
      ctx: GQLContext
    ): Promise<GQLTranscodeJob | GQLPlaybackError> {
      const { id: localVideoId } = fromGlobalId(videoId);
      const internalResolution = gqlResolutionToInternal(resolution);

      // Defensive try/catch — startTranscodeJob now returns its own typed
      // errors for known cases. Anything that still throws is genuinely
      // unexpected (DB connection failure, etc.) and gets mapped to INTERNAL
      // so the client never sees an untyped GraphQL `errors[]` for the
      // playback path. Without this wrapper a single unexpected throw still
      // becomes Relay's "No data returned" protocol violation.
      try {
        const result = await startTranscodeJob(
          localVideoId,
          internalResolution,
          startTimeSeconds,
          endTimeSeconds,
          ctx.otelCtx
        );
        return result.kind === "ok"
          ? presentJob(result.job)
          : presentPlaybackError({
              code: result.code,
              message: result.message,
              retryable: result.retryable,
              retryAfterMs: result.retryAfterMs,
            });
      } catch (err) {
        return presentPlaybackError({
          code: "INTERNAL",
          message: (err as Error).message,
          retryable: false,
        });
      }
    },

    async createLibrary(
      _: unknown,
      {
        name,
        path,
        mediaType,
        extensions,
      }: { name: string; path: string; mediaType: string; extensions: string[] }
    ): Promise<GQLLibrary> {
      const internalType = gqlMediaTypeToInternal(mediaType) as MediaType;
      // createLibrary is a synchronous SQLite INSERT — fast for a local dev store.
      const row = createLibrary(name, path, internalType, extensions);
      // Fire-and-forget background scan so the mutation resolves without waiting.
      void scanLibraries();
      return presentLibrary(row);
    },

    deleteLibrary(_: unknown, { id }: { id: string }): boolean {
      const { id: localId } = fromGlobalId(id);
      return deleteLibrary(localId);
    },

    updateLibrary(
      _: unknown,
      {
        id,
        name,
        path,
        mediaType,
        extensions,
      }: { id: string; name?: string; path?: string; mediaType?: string; extensions?: string[] }
    ): GQLLibrary {
      const { id: localId } = fromGlobalId(id);
      const updated = updateLibrary(localId, {
        name,
        path,
        mediaType: mediaType ? gqlMediaTypeToInternal(mediaType) : undefined,
        extensions,
      });
      if (!updated) throw new Error("Library not found");
      return presentLibrary(updated);
    },

    async matchVideo(
      _: unknown,
      { videoId, imdbId }: { videoId: string; imdbId: string }
    ): Promise<GQLVideo> {
      if (!isOmdbConfigured()) {
        throw new Error("OMDB_API_KEY not configured — go to Settings → Metadata to add it");
      }
      const { id: localVideoId } = fromGlobalId(videoId);
      const video = getVideoById(localVideoId);
      if (!video) throw new Error(`Video not found: ${videoId}`);

      const result = await fetchOmdbById(imdbId);
      if (!result) throw new Error(`OMDb returned no result for IMDb ID: ${imdbId}`);

      const metadata: VideoMetadataRow = {
        video_id: localVideoId,
        imdb_id: result.imdbId,
        title: result.title,
        year: result.year,
        genre: result.genre,
        director: result.director,
        cast_list: result.actors.length > 0 ? JSON.stringify(result.actors) : null,
        rating: result.imdbRating,
        plot: result.plot,
        poster_url: result.posterUrl,
        matched_at: new Date().toISOString(),
      };
      upsertVideoMetadata(metadata);

      return presentVideo(video, true);
    },

    unmatchVideo(_: unknown, { videoId }: { videoId: string }): GQLVideo {
      const { id: localVideoId } = fromGlobalId(videoId);
      const video = getVideoById(localVideoId);
      if (!video) throw new Error(`Video not found: ${videoId}`);
      deleteVideoMetadata(localVideoId);
      return presentVideo(video, false);
    },

    addToWatchlist(_: unknown, { videoId }: { videoId: string }): GQLWatchlistItem {
      const { id: localVideoId } = fromGlobalId(videoId);
      const row = addWatchlistItem(localVideoId);
      return presentWatchlistItem(row);
    },

    removeFromWatchlist(_: unknown, { id }: { id: string }): boolean {
      const { id: localId } = fromGlobalId(id);
      return removeWatchlistItem(localId);
    },

    updateWatchProgress(
      _: unknown,
      { videoId, progressSeconds }: { videoId: string; progressSeconds: number }
    ): GQLWatchlistItem {
      const { id: localVideoId } = fromGlobalId(videoId);
      const row = updateWatchlistProgress(localVideoId, progressSeconds);
      if (!row) throw new Error(`No watchlist item found for video: ${videoId}`);
      return presentWatchlistItem(row);
    },

    setSetting(_: unknown, { key, value }: { key: string; value: string }): boolean {
      setSetting(key, value);
      return true;
    },

    recordPlaybackSession(
      _: unknown,
      { traceId, videoId, resolution }: { traceId: string; videoId: string; resolution: string }
    ): GQLPlaybackSession {
      const { id: localVideoId } = fromGlobalId(videoId);
      const video = getVideoById(localVideoId);
      const row = {
        id: randomUUID(),
        trace_id: traceId,
        video_id: localVideoId,
        video_title: video ? (video.title ?? video.filename) : "Unknown",
        resolution: gqlResolutionToInternal(resolution),
        started_at: new Date().toISOString(),
      };
      insertPlaybackSession(row);
      return presentPlaybackSession(row);
    },
  },
};

File: server-rust/src/graphql/mutation.rs
//! Root Mutation — CRUD for libraries, metadata, watchlist, settings, playback history.

use async_graphql::{Context, Object, ID};

use crate::db::{
    self, add_watchlist_item, create_library, delete_library, delete_video_metadata,
    get_video_by_id, insert_playback_session, remove_watchlist_item, set_setting, update_library,
    update_watchlist_progress, upsert_video_metadata, Db, LibraryUpdate, PlaybackHistoryRow,
    VideoMetadataRow,
};
use crate::graphql::scalars::{MediaType, Resolution};
use crate::graphql::types::{
    Library, PlaybackError, PlaybackSession, StartTranscodeResult, Video, WatchlistItem,
};
use crate::relay::from_global_id;

#[derive(Default)]
pub struct Mutation;

#[Object]
impl Mutation {
    /// Spawn a fire-and-forget background scan, then return the current
    /// library list immediately. Progress flows through the
    /// `library_scan_progress` subscription; the mutation contract is
    /// "kicked off, here are the current libraries". `ScanState::mark_started`
    /// inside `scan_libraries` is the dedup guard, so two concurrent
    /// callers won't both walk the filesystem.
    async fn scan_libraries(&self, ctx: &Context<'_>) -> async_graphql::Result<Vec<Library>> {
        let app_ctx = ctx.data_unchecked::<crate::config::AppContext>();
        let spawn_ctx = app_ctx.clone();
        tokio::spawn(async move {
            crate::services::library_scanner::scan_libraries(&spawn_ctx).await;
        });
        Ok(db::get_all_libraries(&app_ctx.db)?
            .iter()
            .map(Library::from_row)
            .collect())
    }

    /// Cancel a list of in-flight transcode jobs by their global IDs.
    /// Fire-and-forget — the resolver always returns `true` and any
    /// per-id failure is logged inside `FfmpegPool::kill_job`. The
    /// pool's permit-drop runs synchronously, so a follow-up
    /// `start_transcode` from the same client (e.g. the seek path's new
    /// foreground chunk) sees the freed slot in <50 ms instead of
    /// queueing behind the kernel's process reap.
    async fn cancel_transcode(
        &self,
        ctx: &Context<'_>,
        job_ids: Vec<ID>,
    ) -> async_graphql::Result<bool> {
        use crate::services::kill_reason::KillReason;
        let app_ctx = ctx.data_unchecked::<crate::config::AppContext>();
        for global_id in job_ids {
            let (_, local_id) = from_global_id(&global_id)?;
            app_ctx.pool.kill_job(&local_id, KillReason::ClientCancel);
        }
        Ok(true)
    }

    /// Spawn (or reuse) a transcode job for the requested video range.
    /// Returns either a `TranscodeJob` payload (the job is now in the
    /// chunker's job_store and segments will arrive via `/stream/:jobId`)
    /// or a typed `PlaybackError` for the failure modes the client can
    /// react to (capacity exhausted, video not found, probe failed, etc.).
    async fn start_transcode(
        &self,
        ctx: &Context<'_>,
        video_id: ID,
        resolution: Resolution,
        start_time_seconds: Option<f64>,
        end_time_seconds: Option<f64>,
    ) -> async_graphql::Result<StartTranscodeResult> {
        use crate::graphql::types::TranscodeJob;
        use crate::services::chunker::{start_transcode_job, StartJobResult};
        let app_ctx = ctx.data_unchecked::<crate::config::AppContext>();
        let (_, local_id) = from_global_id(&video_id)?;
        let result = start_transcode_job(
            app_ctx,
            &local_id,
            resolution,
            start_time_seconds,
            end_time_seconds,
        )
        .await;
        Ok(match result {
            StartJobResult::Ok(job) => {
                StartTranscodeResult::TranscodeJob(TranscodeJob::from_active(&job))
            }
            StartJobResult::Error {
                code,
                message,
                retryable,
                retry_after_ms,
            } => StartTranscodeResult::PlaybackError(PlaybackError {
                code,
                message,
                retryable,
                retry_after_ms: retry_after_ms.map(|n| n as i32),
            }),
        })
    }

    async fn create_library(
        &self,
        ctx: &Context<'_>,
        name: String,
        path: String,
        media_type: MediaType,
        extensions: Vec<String>,
    ) -> async_graphql::Result<Library> {
        let app_ctx = ctx.data_unchecked::<crate::config::AppContext>();
        let row = create_library(
            &app_ctx.db,
            &name,
            &path,
            media_type.to_internal(),
            &extensions,
        )?;
        // Fire-and-forget background scan so a freshly-added profile gets
        // indexed without the user having to click "Scan All".
        let spawn_ctx = app_ctx.clone();
        tokio::spawn(async move {
            crate::services::library_scanner::scan_libraries(&spawn_ctx).await;
        });
        Ok(Library::from_row(&row))
    }

    async fn delete_library(&self, ctx: &Context<'_>, id: ID) -> async_graphql::Result<bool> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&id)?;
        Ok(delete_library(db, &local_id)?)
    }

    async fn update_library(
        &self,
        ctx: &Context<'_>,
        id: ID,
        name: Option<String>,
        path: Option<String>,
        media_type: Option<MediaType>,
        extensions: Option<Vec<String>>,
    ) -> async_graphql::Result<Library> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&id)?;
        let updated = update_library(
            db,
            &local_id,
            LibraryUpdate {
                name: name.as_deref(),
                path: path.as_deref(),
                media_type: media_type.map(|m| m.to_internal()),
                extensions,
            },
        )?
        .ok_or_else(|| async_graphql::Error::new("Library not found"))?;
        Ok(Library::from_row(&updated))
    }

    /// Step 1 stub — writes an empty metadata row keyed by IMDb ID. The full
    /// OMDb fetch ships in Step 2 alongside the scanner.
    async fn match_video(
        &self,
        ctx: &Context<'_>,
        video_id: ID,
        imdb_id: String,
    ) -> async_graphql::Result<Video> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&video_id)?;
        let video = get_video_by_id(db, &local_id)?
            .ok_or_else(|| async_graphql::Error::new(format!("Video not found: {video_id:?}")))?;
        let now = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        upsert_video_metadata(
            db,
            &VideoMetadataRow {
                video_id: local_id.clone(),
                imdb_id: imdb_id.clone(),
                title: video
                    .title
                    .clone()
                    .unwrap_or_else(|| video.filename.clone()),
                year: None,
                genre: None,
                director: None,
                cast_list: None,
                rating: None,
                plot: None,
                poster_url: None,
                poster_local_path: None,
                matched_at: now,
            },
        )?;
        Ok(Video::from_row(&video))
    }

    async fn unmatch_video(&self, ctx: &Context<'_>, video_id: ID) -> async_graphql::Result<Video> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&video_id)?;
        let video = get_video_by_id(db, &local_id)?
            .ok_or_else(|| async_graphql::Error::new(format!("Video not found: {video_id:?}")))?;
        delete_video_metadata(db, &local_id)?;
        Ok(Video::from_row(&video))
    }

    async fn add_film_to_watchlist(
        &self,
        ctx: &Context<'_>,
        film_id: ID,
    ) -> async_graphql::Result<WatchlistItem> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&film_id)?;
        let row = add_watchlist_item(db, &local_id)?;
        Ok(WatchlistItem::from_row(&row))
    }

    async fn remove_from_watchlist(
        &self,
        ctx: &Context<'_>,
        id: ID,
    ) -> async_graphql::Result<bool> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&id)?;
        Ok(remove_watchlist_item(db, &local_id)?)
    }

    async fn update_watch_progress(
        &self,
        ctx: &Context<'_>,
        film_id: ID,
        progress_seconds: f64,
    ) -> async_graphql::Result<WatchlistItem> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&film_id)?;
        let row = update_watchlist_progress(db, &local_id, progress_seconds)?
            .ok_or_else(|| async_graphql::Error::new("No watchlist item for film"))?;
        Ok(WatchlistItem::from_row(&row))
    }

    async fn set_setting(
        &self,
        ctx: &Context<'_>,
        key: String,
        value: String,
    ) -> async_graphql::Result<bool> {
        let db = ctx.data_unchecked::<Db>();
        set_setting(db, &key, &value)?;
        Ok(true)
    }

    async fn record_playback_session(
        &self,
        ctx: &Context<'_>,
        trace_id: String,
        video_id: ID,
        resolution: Resolution,
    ) -> async_graphql::Result<PlaybackSession> {
        let db = ctx.data_unchecked::<Db>();
        let (_, local_id) = from_global_id(&video_id)?;
        let video = get_video_by_id(db, &local_id)?;
        let video_title = video
            .map(|v| v.title.unwrap_or(v.filename))
            .unwrap_or_else(|| "Unknown".into());
        let id = uuid::Uuid::new_v4().to_string();
        let started_at = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let row = PlaybackHistoryRow {
            id: id.clone(),
            trace_id: trace_id.clone(),
            video_id: local_id,
            video_title: video_title.clone(),
            resolution: resolution.to_internal().into(),
            started_at: started_at.clone(),
        };
        insert_playback_session(db, &row)?;
        Ok(PlaybackSession {
            id: ID(id),
            trace_id,
            video_title,
            resolution,
            started_at,
        })
    }
}
